### DISCUSSION TITLE
How does TinyUSB process the descriptors?

### DISCUSSION THREAD

**Discussion by nachiketathakur** [hathach/tinyusb#1386] [Q&A] [answered]
I am going through the dev_hid_composite example and I cannot understand how the descriptors are passed? In the example the piece of code which changes the descriptor is as follows: 
```
uint8_t const desc_configuration[] =
{
  // Config number, interface count, string index, total length, attribute, power in mA
  TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 100),

  // Interface number, string index, protocol, report descriptor len, EP In address, size & polling interval
  TUD_HID_DESCRIPTOR(ITF_NUM_HID, 0, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report), EPNUM_HID, CFG_TUD_HID_EP_BUFSIZE, 5)
}
```
In the line: ``` TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 100), ```
There are total 6 arguments which are passed and if we refer to the https://www.beyondlogic.org/usbnutshell/usb5.shtml there are total 8 fields of configuration descriptor. What I don't understand is how does it works and how can I add my own configuration descriptors of my project into a project which uses TinyUSB device?

**Comment by hathach** [comment, marked as answer]
it is template using macro, check out its definition for details

https://github.com/hathach/tinyusb/blob/73896a3b71c525a3ee4cefa7e35ce3b3a93786ef/src/device/usbd.h#L187-L188

https://github.com/hathach/tinyusb/blob/73896a3b71c525a3ee4cefa7e35ce3b3a93786ef/src/device/usbd.h#L247-L253

**Reply by nachiketathakur** [reply]
Thank you for the quick reply. I still do not understand why only 6 out of the 8 arguments are passed to the function macro. Also, the ```\``` in the function macro... I understand that it is simply tying the part after ```\``` with the function macro, but how are those lines implemented in the function macro? 
Are they the default values to the argument of function macros?

**Reply by hathach** [reply]
it is just pure C macro expansion, nothing special.
